### DISCUSSION TITLE
Elfin EW11a logging CRC errors.

### DISCUSSION THREAD

**Discussion by LevittS** [nathanmarlor/foxess_modbus#339] [General]
Hi,

Having fought quite hard with the fragility of the CT and Cat5 cable I believe I've got the connector from my H1 successfully wired and can see received and send bytes in the serial port state of the EW11.

However, I'm not getting any responses from the EW11s TcpServer.

Adding a logging endpoint I can see what's stopping the responses being returned, I believe:
```
[0-0:26:37]<tcpd> recv:12-byte
[0-0:26:37]<tcpd> recv_byte=360,recv_frame=30
[0-0:26:37]<uart> send byte:8-byte,failed:0-byte
[0-0:26:37]<uart> send_byte=296,send_frame=30
[0-0:26:37]<uart> failed_byte=0,failed_frame=0
[0-0:26:37]<process> modbus from(192.168.0.148,41156):transId=02,protoId=00

[0-0:26:37]<uart> recv:6-byte
[0-0:26:37]<uart> recv_byte=230,recv_frame=30
[0-0:26:37]<process> modbus read:wrong CRC!

```
Is this definitely indicative of a cabling issue?
Or could it be something like the firmware version of my inverter?

If I change the protocol on the serial settings from modbus to none and send modbus packets I can see that the raw data being returned from the inverter changes each time - but not sure if that would be expected from what I assume is some kind of 10 byte error response.

I've built my own RJ45 cable using two twisted pairs for the power and 485 data in a cat5 cable - the power being then patched onto a USB-A terminated cable.
I have the cat5 -> usb patch connection within the inverter plug (as it makes for a reasonably tidy solution) - could that be causing noise affecting the data comms?

Just trying to work out if there's anything I can do without mucking around within the inverter plug (again) and/or whether its a reasonable assumption that separating the power and data lines outside of the plug would make any difference?

Thanks

**Comment by iancg** [comment]
I have both my CT clamp and Modbus in the same cat 5 cable.  The installer had originally twisted two pairs together, but in an odd way, i.e. they had twisted brown and brown stripe as +ve and orange and orange stripe as -ve, i.e. in effect negating the benefit of it being a twisted pair and increasing the risk of cross-talk.

Because I wanted to use the extra pair in the existing cable rather than run a new cable (I have two inverters sharing the same cable), I separated them and used brown as +ve and brown stripe as -ve for the CT clamp and orange as A and orange stripe as B for modbus.

The other inverter uses the other two twisted pairs.

The cable length is about 5m and I have no errors (how long is your cable from the EW11 to inverter?)

I can't see why having USB power (should be a clean 5v DC) in the same cable should cause any cross talk.

Just to confirm, have you configured the EW11 as baud rate 9600, parity None and stopbits 1?

The error you are getting suggests the issue is between the EW11 and inverter (see https://control.com/forums/threads/crc-checking-error-in-modbus-protocol.16008/ if you are interested in the detail of how modbus CRCs work).

Sorry - I don't have a direct answer, but hopefully some of my own experience and ramblings help.

**Reply by LevittS** [reply]
Cable is around 50cm.
Yep - 9600 8N1

**Reply by LevittS** [reply]
Maybe I'll try redoing the RJ45 end of my cable - perhaps I've got a dodgy connection at that end - at least that would not require opening the inverter end.

Thanks for your thoughts.

**Reply by canton7** [reply]
Another thing you  can try is putting it in serial mode (no modbus), then either using modpoll with `-m enc` or add it to Home Assistant as "Other -> Modbus over TCP". That might get you a different / more descriptive error message.

**Reply by LevittS** [reply]
So, redid the RJ45 connector - has changed nothing.

Putting the Serial protocol to None, and changing my modbus client code to use RTU framing I can see the 8 byte request and 6 byte response in wireshark.

Request:
`f7 03 75 30 00 1c 4a 96`

Response:
`04 7e 45 f6 c7 6b`

The response doesn't vary at all - when I said it was earlier I was seeing the TCP header changes.

The request is attempting to read the model and version information that works on the direct ethernet connection.

**Reply by canton7** [reply]
That request looks fine, but the response looks like garbage as far as I can tell. It doesn't start with the slave ID, doesn't contain the function code `03` or the error code `83`, and is the wrong length. I'm really not sure what's going on.

I'd try using a dedicated wire for RS485, keep it short, keep it separate to everything else. Just see whether that works, and if it does you can take things from there.

**Reply by FozzieUK** [reply]
I’ve seen similar framing and CRC errors when you connect the device to the CT connector meterRS485 pins 1&2 (master) instead of the RS485 (slave) on pins 3 & 4 - it’s an easy mistake to make so worth double checking you are on pins 3 & 4.

**Reply by LevittS** [reply]
I couldn't be much more certain I'm using pins 3 & 4  - the connected pins are not on the same 'level' of the plug.

I had to remove pins 1 & 2 from the plug as I found out the hard way they are only held in by the grub screws due to me stupidly managing to drop two of them somewhere in my loft.

Surely that can't of made pins 3 &4 take on the function of 1 & 2 (of which A & B are actually reversed)?

I also ended up with the wires connected the other way around than I thought they should be (if A -> A and B -> B) - 
which somehow correlates.

**Reply by canton7** [reply]
What do you mean by "not on the same level"? They should all be identical iirc.

**Reply by canton7** [reply]
And yes, A connects to A. Beware than pin 4 is A, and 3 is B.

**Reply by LevittS** [reply]
In the plug you have:
```
3 2 1
 5 4
8 7 6
```
So 4 is on a different 'level' to 3 - sorry - couldn't think of a better term.

**Reply by FozzieUK** [reply]
Ok, as I said it was worth asking I’ve seen that done a few times with similar crc/ frame errors.
![A3EDD8EC-1774-43FE-9ECB-30B41218B24A](https://github.com/nathanmarlor/foxess_modbus/assets/113460294/32eae0b2-3a5a-461b-a640-98ec0251e870)

**Reply by dreampi** [reply]
@LevittS, could you upload a picture of your plug with the wiring to help us understand?

**Reply by LevittS** [reply]
@dreampi, Not without taking it all apart again, and nearly every time I do that I end up breaking one of the wires.

**Reply by LevittS** [reply]
Llike I say the EW11 didn't receive anything, I believe, until I swapped the 3 & 4 wires around from what I thought they should be; I spent quite some time trying to make sure I'd got it sorted and written down before I started, unless of course I've double bluffed myself.

I think I might have to find a way to get the unconnected 1 & 2 pins back into the plug as that might not be helping, and is probably different from anything else anyone has.

I did read somewhere about the EW11 not working for someone until they hard/factory reset it - wondering whether that is worth a go before I start ripping it all apart - but that gave the impression it wasn't working at all, rather than giving odd responses.

If none of that helps, then I'll redo it with a single twisted pair exiting the plug and do the USB power splicing elsewhere.

Thanks all for all of your input, it's given me some avenues to explore.

**Reply by LevittS** [reply]
I managed to find the two grub screws I'd dropped with a magnet and have put pins 1 & 2 back into the plug (so I know for sure that I'm connecting to pins 3 &4), and perhaps understandably and thankfully, that didn't make any difference - still getting the same results as before.

Took a pictures of the EW11s RJ45 and the 485 pin connections - and I've definitely got the EW11's pin 5 (solid blue) connected to the inverters pin 3 - which I'm pretty certain is wrong, but until I did that I got absolutely nothing.

![IMG_20230624_095441](https://github.com/nathanmarlor/foxess_modbus/assets/16302343/b199d0c9-7f26-4234-a6e2-476e786fef27)
![IMG_20230624_095229](https://github.com/nathanmarlor/foxess_modbus/assets/16302343/d5d93102-e4e1-4bc6-83a3-40ee3b529a90)

I'm going to make up a 'reset' cable for the EW11 and also a patch cable that swaps the blue/white and solid blue cables so I can experiment without having to turn off my inverter (again).

**Reply by FozzieUK** [reply]
well done on finding the screws, i'm never that lucky :+1

Looking at the Elfin manual, as you say, you are definitely connected to the wrong pins - the patch cable is a good idea.

Did you try the factory reset?, a few comments suggest that has resolved 'issues' with comms

**Reply by LevittS** [reply]
I've tried a reset via the web-ui but it didn't make any difference - I'm going to make up a separate reset cable and try that (connecting pin 4 to ground).

I shall do that once I get some more rj45 plugs, I'd ordered some off ebay a few weeks ago, and the seller seems to have let me down, so just ordered some more from elsewhere along with a coupler - So hopefully will get those Tue/Wed next week.

:crossed_fingers:

**Reply by LevittS** [reply]
RJ45s just arrived.

Threw together a patch cable - and thought I'd try that before resetting the EW11 - and it WORKS.

So the CRC / Framing errors can also occur if you've got the data wires the wrong way round it seems.

So, somewhat annoyingly, despite a couple of attempts of connecting and checking the wiring originally, I must have got something wrong (perhaps I tightened the grub screw onto the cable sheath repeatedly) until I swapped the data wires, or the EW11 needed some kind of 'kick' to get it sending stuff properly (I didn't software reset it until after I started getting the bad data back, so might not be that; and I've not performed a pin 4 grounding reset - didn't actually build the reset cable).

Kind of tempted to redo the RJ45 connector end now so I don't need the coupler and patch cable, but also tempted to leave well alone as its working....

Thanks to all of you for the fast and informative responses, it gave me the energy to keep going.

Now onto using the additional data I should have access to, and the less reliance on FoxEss' 'portal' that will afford.

Cheers!
